lib/compile.js

```diff
@@ -36,29 +36,56 @@ function compileLoad(loader, load, opts, compilers) {
   });
 }
 
+function retrieveCachedOutput(cache, load) {
+  if (! cache)
+    return;
+
+  var entry = cache[load.name];
+  if (! entry)
+    return;
+
+  return load.source === entry.source && entry.output;
+}
+
+function updateCacheEntry(cache, load, output) {
+  if (! cache)
+    return;
+
+  cache[load.name] = {
+    source: load.source,
+    name: load.name,
+    output: output,
+  };
+}
+
 exports.compileOutputs = compileOutputs;
-function compileOutputs(loader, tree, opts, sfxEntryPoints) {
+function compileOutputs(loader, tree, opts, sfxEntryPoints, cache) {
   var modules = Object.keys(tree);
 
   // store plugins with a bundle hook to allow post-processing
   var plugins = {};
   // store compiler list for sfx bundling
   var compilers = {};
-  var outputs = [];
 
   var usesBabelHelpersGlobal, usesTraceurRuntimeGlobal;
 
+  // to pass list of outputs through promise chain.
+  var outputs;
+
   return Promise.all(modules.map(function(name) {
     var load = tree[name];
 
     if (load === null)
       throw new TypeError('"' + name + '" was defined via a bundle, so can only be used for subtraction or union operations.');
 
-    if (sfxEntryPoints && load.metadata.loader && load.metadata.build === false)
-      outputs.push('System.registerDynamic("' + load.name + '", [], false, function() { console.log("SystemJS Builder - Plugin for ' + load.name + ' does not support sfx builds"); });\n');
-
-    if (load.metadata.build === false)
-      return;
+    if (load.metadata.build === false) {
+      if (sfxEntryPoints && load.metadata.loader) {
+        return 'System.registerDynamic("' + load.name + '", [], false, function() { console.log("SystemJS Builder - Plugin for ' + load.name + ' does not support sfx builds"); });\n';
```

Comment:
If you follow the promise chain you'll see it's equivalent and, IMO, a little neater.